8333268: Fixes for static build

magicus opened the pull request:
This patch contains a set of changes to improve static builds. They will pave the way for implementing a full static-only java launcher. The changes here will:

1) Make sure non-exported symbols are made local in the static libraries. This means that the risk of symbol conflict is the same for static libraries as for dynamic libraries (i.e. in practice zero, as long as a consistent naming scheme is used for exported functions).

2) Remove the work-arounds to exclude duplicated symbols.

3) Fix some code in hotspot and the JDK libraries that did not work properly with a static java launcher.

The latter fixes are copied from or inspired by the work done by @jianglizhou and her team as part of the Project Leyden [Hermetic Java](https://github.com/openjdk/leyden/tree/hermetic-java-runtime).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8333268](https://bugs.openjdk.org/browse/JDK-8333268): Fixes for static build (**Bug** - P3)


### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) 🔄 Re-review required (review applies to [b88d813e](https://git.openjdk.org/jdk/pull/19478/files/b88d813e33c23154911ca5550da18eec4a2cae5c))

### Contributors
 * Jiangli Zhou `<jiangli@openjdk.org>`
 * Magnus Ihse Bursie `<ihse@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19478/head:pull/19478` \
`$ git checkout pull/19478`

Update a local copy of the PR: \
`$ git checkout pull/19478` \
`$ git pull https://git.openjdk.org/jdk.git pull/19478/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19478`

View PR using the GUI difftool: \
`$ git pr show -t 19478`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19478.diff">https://git.openjdk.org/jdk/pull/19478.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19478#issuecomment-2168638632)

magicus commented:
Some open questions:

* Do `os::lookup_function` need to be implemented on Windows too, for symmetry, even if it is only used on Unix platforms?

* Many of the changes in Hotspot boils down to `os::dll_load` doing the wrong thing when running with a static build. Perhaps we should provide a better function that knows how to find and load a symbol for both static and dynamic builds, and use that instead of making a lot of tests for static/dynamic on each location we need to look up a symbol from some other JDK library.

* I managed to replace most of the #ifdef STATIC_BUILD with runtime checks. There are some places remaining though. Apart from the #ifdefs needed for JNI/JVMTI, which will need spec changes to address, there are code in java_md_macosx.m, jio.c and awt_Mlib.c that I did not manage to turn into runtime checks. They will need some more thorough work than just changing an `#ifdef` to an `if () {`.

* And of course, the code in the build system to share all .o files except the two linktype files is still under development...

magicus commented:
I moved this away from Draft state, since I think it needs some visibility, especially since it touches several different parts of the code base, and such reviews tend to take time.

I think the code here is good and basically okay to integrate. This patch will not on it's own solve the entire problem of building a proper static launcher, but it takes several important steps along the way. I think the changes here are reasonable to integrate into mainline at this point.

magicus commented:
The GHA tests fails when building gtest on Linux. This will require some investigation.

magicus commented:
> Do os::lookup_function need to be implemented on Windows too, for symmetry, even if it is only used on Unix platforms?

@AlanBateman suggested to add `lookup_function` to os_windows.cpp as well, but just let it contain ShouldNotReachHere. That sounds like a good solution to me.

magicus commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -585,22 +585,37 @@ void os::init_system_properties_values() {
+      }
+
+      // Get rid of /{client|server|hotspot}, if binary is libjvm.so.
+      // Or, cut off /<binary_name>.
```
@jianglizhou This code is based on changes in the Hermetic Java repo, but I do not fully understand neither the comment nor what the purpose is. If you could explain this a bit I'd be grateful.

jianglizhou replied:
The specific related commit in the hermetic Java branch is https://github.com/openjdk/leyden/commit/53aa8f0cf418ab5f435a4b9996c7754fb8505d4b. 

The change in os_linux.cpp here is to make sure  that the libjvm.so related path manipulation is conditionally done only. The check at line 599 looks for "/libjvm.so" substring, so we only chop off (`*pslash = `\0` at line 601) that part when necessary. In the static JDK case, there is no `libjvm.so` and the path string is `<jdk_path>/bin/javastatic`, which should not be affected. Otherwise, it could fail.

I found the code was not very easy to follow when running into problems and fixing for static support. So I added a bit more comments in the code here. The comment above about `/{client|server|hotspot}` was there originally. I think we no longer have those directories. We can cleanup that later, since it needs some more testing.

@magicus, thanks a lot for extracting/reworking/cleaning-up the static Java changes from the hermetic Java branch. That's a substantial amount of work!

I have one quick comment about the removal of `STATIC_LIB_EXCLUDE_OBJS` changes. Will post it as separate comment for the related code. 

I'll also look closely of the vm & jdk changes and compare those with the changes in the hermetic Java branch this week.

magicus commented:
The reason the gtest failed was that we build a static library libgtest.a, which is linked with the gtest libjvm.so. With the changes in this PR, libgtest.a was being built using the `ld -r` + `objcopy --localize-hidden` method. This caused some weird issues with gcc, related to C++ code and the `COMDAT` object info. 

I failed to track down any proper solution, so instead I added a patch where the libraries that we explicitly declare as `STATIC_LIBRARIES` are linked as before, without the partial linking step. These libraries are only intended for internal consumption (that is, they are linked to and used by another, "external" library), and so the extra protection added by the partial linking is not really needed.

It's a bit sad that this did not work, but it is no big deal. It won't affect files released in the image, and it will not be a regression as compared to now.

magicus commented:
/contributor add jiangli
/contributor add ihse

AlanBateman commented:
The changes to the launcher look okay. The move from `ifdef STATIC_BUILD` to `JLI_IsStaticallyLinked` is quite nice.

Having libjdwp link to libjvm was a surprise but I think okay.

I think it would be useful to provide a brief summary on the when/where the static builds are tested to ensure that the changes don't bit rot. I realise we already have static builds but it isn't obvious where this is tested.

jianglizhou commented on make/modules/java.base/lib/CoreLibraries.gmk:
```diff
@@ -144,14 +144,6 @@ ifneq ($(USE_EXTERNAL_LIBZ), true)
       $(addprefix $(TOPDIR)/src/java.base/share/native/libzip/zlib/, \
           $(LIBJLI_EXTRA_FILE_LIST))
-
-  # Do not include these libz objects in the static libjli library.
```
Why this is no longer needed?

jianglizhou commented on make/modules/java.desktop/lib/AwtLibraries.gmk:
```diff
@@ -273,7 +256,6 @@ ifeq ($(call isTargetOs, windows macosx)+$(ENABLE_HEADLESS_ONLY), false+false)
       LIBS_unix := $(LIBDL) $(LIBM) $(X_LIBS) -lX11 -lXext -lXi -lXrender \
           -lXtst, \
```
Same question as for the STATIC_LIB_EXCLUDE_OBJS change with `libjli`. Are the duplicate symbol issues resolved by symbol hiding?

I think it's still better to not include those .o files to avoid unnecessary footprint overhead in the binary.

magicus replied:
> Are the duplicate symbol issues resolved by symbol hiding?

Yes, correct.

> I think it's still better to not include those .o files to avoid unnecessary footprint overhead in the binary.

This was introduced as a being completely necessary to produce static builds, not as a means to reduce binary size. These kind of hard-coded lists are always a maintenance problem, and I do not think they are worth keeping just for the sake of reducing binary size.

If that is an issue, we can probably add some LTO step when doing the partial link. I'd guess that will be much more beneficial in terms of size reduction that to exclude a handful of manually picked files.

Having this kind of excludes furthermore makes it harder to selectively only include libraries belonging to a specific subset of modules, since it means that some libraries will not work unless other -- unspecified -- libraries are also included in the link.

jianglizhou commented on make/common/native/Link.gmk:
```diff
@@ -69,10 +69,15 @@ endef
-    ifneq ($(findstring $(TOOLCHAIN_TYPE), clang), )
-      $1_ENABLE_PARTIAL_LINKING := true
-    endif
+  ifeq ($(call isTargetOs, linux macosx), true)
```
Is this mainly for `clang` support for now?

magicus replied:
I'm not sure I understand what you mean. I have been continually testing clang on macOS and linux, and gcc on linux, while developing this patch. Since these are the only supported compilers on these platforms, I have a platform-specific test rather than a compiler-specific test.

jianglizhou commented on src/hotspot/os/bsd/os_bsd.cpp:
The changes in os_bsd.cpp are new and are not from https://github.com/openjdk/leyden/tree/hermetic-java-runtime/. Have you tested the bsd port?

magicus replied:
bsd in Hotspot means "macOS" in practice. I have tested my changes on macOS. The mainline of OpenJDK does not support any other BSDs.

jianglizhou commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -518,6 +518,11 @@ void* os::native_java_library() {
     char buffer[JVM_MAXPATHLEN];
     char ebuf[1024];
 
+    if (vm_is_statically_linked()) {
```
This block can be moved before the two variable declarations above, since they are not needed in the static case. https://github.com/openjdk/leyden/blob/c1c5fc686c1452550e1b3663a320fba652248505/src/hotspot/share/runtime/os.cpp#L507 handles it that way.

magicus replied:
Fixed.

jianglizhou commented on src/hotspot/share/utilities/zipLibrary.cpp:
```diff
@@ -60,6 +60,10 @@ static inline bool not_loaded() {
 }
 
 static void* dll_lookup(const char* name, const char* path, bool vm_exit_on_failure) {
+  if (vm_is_statically_linked()) {
```
I like this change. It is cleaner than the hermetic Java branch change that does the `if` static check in `store_function_pointers` (https://github.com/openjdk/leyden/blob/c1c5fc686c1452550e1b3663a320fba652248505/src/hotspot/share/utilities/zipLibrary.cpp#L75).

jianglizhou commented on src/java.base/macosx/native/libjli/java_md_macosx.m:
In the mailing list email discussion thread on hermetic Java, you mentioned running on macosx with a build from hermtic Java branch crashed for you during startup. Is that fully resolved with the changes in this PR? The hermetic Java branch does not have any changes for macosx port. What tests are done for the macosx port for static support?

magicus replied:
Yes, that is fully resolved, although the code I have used to test the macOS port is not included. I'll make sure that there is a proper way to verify the static build on macOS when I've added the makefile machinery for creating a proper static executable.

jianglizhou commented on src/java.base/unix/native/libjli/java_md.c:
```diff
@@ -313,31 +313,34 @@ CreateExecutionEnvironment(int *pargc, char ***pargv,
-        JLI_ReportErrorMessage(CFG_ERROR7);
-        exit(1);
-    }
+    if (!JLI_IsStaticallyLinked()) {
```
Any reason this is diverted from the change in hermetic Java branch, https://github.com/openjdk/leyden/blob/c1c5fc686c1452550e1b3663a320fba652248505/src/java.base/unix/native/libjli/java_md.c#L300? I think the setenv part below is not needed for static/hermetic support either. There is no $JRE/lib with a single executable image. All natives are statically linked with the executable image.

magicus replied:
I do not recall my reason for deviating from the hermetic Java branch. I have double-checked the code, and I agree with you that the setenv part is not needed for static libraries, since all it does is trying to use `LD_LIBRARY_PATH` properly to load the dynamic libraries, so the short-cut solution from the hermetic-java-runtime branch would suffice.

However, the `SetExecname()` call must be done, otherwise the re-execution of the binary as done on macOS (to be able to release the main thread to the GUI handler) will fail. This was probably the reason I started changing this function at all. But after the `SetExecname()` call, we can exit the rest if we are on a static build.

jianglizhou reviewed (commented):
I've looked through all JDK and VM changes and left comments in various places. All the rest changes in PR look good. Thanks again for extracting these changes from the leyden/hermeticJava branch and integrating with mainline!

My other main question is why the `javastatic` linking work is not included in the PR together with these runtime changes. 

IIUC from our meetings and mailing list discussions, the initial integration PR needs to include the part for statically linking the `javastatic`. That's a minimum requirement for testing/verifying the runtime changes when integrating into the mainline, which is also the reason why we haven't starting integrating any of the runtime changes so far. Has that been changed?

magicus commented:
I've reverted this PR back to Draft state. This was how I originally intended for this to be viewed at this stage, but I got struck by pre-vacation hubris and thought that maybe it was after all ready for integration, so I could get it done with before I was leaving for the summer. This was clearly not the case.

I'll look through all comments to this PR carefully and determine how to go forward with this. It might be that this PR still tries to do too many things at once (even though I have removed all the hermetic-java specific changes), so that it needs to be split into several different PRs.

magicus commented on src/hotspot/share/runtime/linkType.cpp:
```diff
@@ -0,0 +1,38 @@
+#ifdef STATIC_BUILD
+  return JNI_TRUE;
+#else
+  return JNI_FALSE;
```
I tried to fit in with the other functions in java.hpp, which all started with `vm_`. Now with a closer look I realize this was perhaps not a prefix shared by all functions. Would you say `is_vm_statically_linked` or just `is_statically_linked` is a better name?

So this should either be made to return `jboolean`, or return `false` or `true`, is that correct? Which one would be preferred? (Sorry for being rusty on this, last time I did any serious VM development was in C for JRockit.)

magicus replied:
I opted for is_vm_statically_linked and to return actual `bool` literals. Let me know if this was the wrong choice.

magicus commented:
> IIUC from our meetings and mailing list discussions, the initial integration PR needs to include the part for statically linking the javastatic. That's a minimum requirement for testing/verifying the runtime changes when integrating into the mainline, which is also the reason why we haven't starting integrating any of the runtime changes so far. Has that been changed?

No, that has not been changed, and it was the "get this done before vacation" stress that got the better of me, and I forgot or ignored (I don't recall which anymore) that this is an absolute requirement for this code to be properly tested. 

I'll get back to fixing the proper static launcher build, and get that properly working, before trying to push any other changes into mainline.

magicus commented:
@AlanBateman 

> Having libjdwp link to libjvm was a surprise but I think okay.

This is since it needs to access JVM_IsStaticallyLinked. The alternative is to create yet another "is static" file that is compiled twice, just for libjdwp, but I think already having two is one too many (even if I realize that it is neccessary), so I'd like to avoid that.

As a rule of thumb, most JDK libraries link with either libjvm or libjava or both, so I don't think this needs to be of any concern.

magicus commented:
I have broken out the parts that is just about replacing `#ifdef`s with a runtime function call to determine static/dynamic loading, into https://github.com/openjdk/jdk/pull/20666. That should hopefully be trivial to integrate since it is a pure refactoring, but it will reduce the size and complexity of this PR.

magicus commented:
This is superseded by https://github.com/openjdk/jdk/pull/20837.

This PR did contain changes that would hide local symbols on more build targets/toolchains than linux/clang; these should be resurrected in a future follow-up on JDK-8339480.
